client/state/themes/actions.js

```diff
@@ -454,3 +460,83 @@ export function clearThemeUpload( siteId ) {
 		siteId,
 	};
 }
+
+export function initiateThemeTransfer( siteId, file ) {
+	return dispatch => {
+		dispatch( {
+			type: THEME_TRANSFER_INITIATE_REQUEST,
+			siteId,
+		} );
+		return wpcom.undocumented().initiateTransfer( siteId, null, file, ( event ) => {
+			dispatch( {
+				type: THEME_TRANSFER_INITIATE_PROGRESS,
+				siteId,
+				loaded: event.loaded,
+				total: event.total,
+			} );
+		} )
+			.then( ( { transfer_id } ) => {
+				dispatch( {
+					type: THEME_TRANSFER_INITIATE_SUCCESS,
+					siteId,
+					transferId: transfer_id,
+				} );
+				themeTransferStatus( siteId, transfer_id )( dispatch );
+			} )
+			.catch( error => {
+				dispatch( {
+					type: THEME_TRANSFER_INITIATE_FAILURE,
+					siteId,
+					error,
+				} );
+			} );
+	};
+}
+
+export function themeTransferStatus( siteId, transferId, interval = 3000, timeout = 180000 ) {
+	const endTime = Date.now() + timeout;
+	return dispatch => {
+		const pollStatus = ( resolve, reject ) => {
+			if ( Date.now() > endTime ) {
+				// timed-out, stop polling
+				dispatch( {
+					type: THEME_TRANSFER_STATUS_FAILURE,
+					siteId,
+					transferId,
+					error: 'client timeout',
+				} );
+				return resolve();
+			}
+			return wpcom.undocumented().transferStatus( siteId, transferId )
+				.then( ( { status, message, themeId } ) => {
+					dispatch( {
+						type: THEME_TRANSFER_STATUS_SUCCESS,
+						siteId,
+						transferId,
+						status,
+						message,
+						themeId,
+					} );
+					if ( status === 'complete' ) {
+						// finished, stop polling
+						return resolve();
+					}
+					// poll again
+					return delay( pollStatus, interval, resolve, reject );
+				} )
+				.catch( ( error ) => {
+					return reject( error );
+				} );
+		};
+
+		const poll = new Promise( pollStatus );
```

Comment:
The only reason we return the Promise is so we can have async tests. We need to not tear-down the test until the async stuff has finished, so that we can check all the dispatches. Only alternative, I think, would be to take a callback, which seems even weirder in production.

Open to other ideas?